client/src/lib/firebase/auth.ts

```diff
@@ -37,14 +40,84 @@ export async function loginAnonymous() {
 
 export function deleteAccount() {
   const user = auth.currentUser;
-
   if (user == null) {
     throw new Error("User is not defined.");
   }
 
   deleteUser(user);
 }
 
+export function linkWithGoogle(user: User | null) {
+  if (user === undefined) {
+    throw new Error("Not signed in");
+  }
+
+  if (user != null) {
+    if (hasGoogleProvider(user)) {
+      throw new Error("google account already linked");
+    }
+    const google = new GoogleAuthProvider();
+    return linkWithPopup(user, google);
+  }
+}
+
+export function linkWithMicrosoft(user: User | null) {
+  if (user === undefined) {
+    throw new Error("Not signed in");
+  }
+
+  if (user != null) {
+    if (hasMicrosoftProvider(user)) {
+      throw new Error("microsoft account already linked");
+    }
+
+    const microsoft = new OAuthProvider("microsoft.com");
+    return linkWithPopup(user, microsoft);
+  }
+}
+
+export function linkWithPassword(password: string) {
+  const user = auth.currentUser;
+  if (user === undefined) {
+    throw new Error("Not Signed In");
+  }
+
+  if (user != null) {
+    return updatePassword(user, password);
+  } else {
+    return user;
+  }
+}
+
+export function hasGoogleProvider(user: User | null) {
+  if (user == null) {
+    return false;
+  }
+
+  const signInMethods = user.providerData;
+
+  for (const provider of signInMethods) {
+    if (provider.providerId == "google.com") {
+      return true;
+    }
+  }
+  return false;
+}
+
+export function hasMicrosoftProvider(user: User | null) {
+  if (user == null) {
+    return false;
+  }
+  const signInMethods = user.providerData;
```

Comment:
you put a blank line here in `hasGoogleProvider` but not in `hasMicrosoftProvider`

Reply:
fixed in f7acc7b53c359a49a8d55d432d7edb0e5a160dee